### PR TITLE
Prevent disappearance of mouse when SpinBox is hidden while dragging

### LINKED ISSUE #76729: AnimatedSprite2d: SpriteFrames Editor
### Godot version

v4.1.dev1.official [db1302637]

### System information

Windows 10, Compatibility Rendering Driver, GTX 1660 Ti (531.79)

### Issue description

The mouse cursor disappears when i'm clicking and holding the mouse button on the little arrows in the Frame Duration option and moving the mouse over the Edit Box (with the x1 text).

image

### Steps to reproduce

1. Add AnimatedSprite2d node
2. Load spritesheet
3. Open SpriteFrames editor
4. Select a frame
5. click and hold the mouse button over the little up or down arrow on the Frame Duration Option
6. While holding the mouse button move left over the edit box.

### Minimal reproduction project

[FrameDurationTest.zip](https://github.com/godotengine/godot/files/11398451/FrameDurationTest.zip)

### LINKED ISSUE #77803: Mouse disappear when SpinBox is being hidden while there is a Drag
### Godot version

4.1.dev (db5113de36)

### System information

macOS 13.3.0 - Vulkan (Forward+) - integrated Apple M1 Pro - Apple M1 Pro (8 Threads)

### Issue description

When hiding a SpinBox during a grab modification. The mouse will stay hidden eternally as SpinBox would stop receiving inputs. Because of how SpinBox is implemented, there is another way to trigger this bug that is shown in MRP.

This issue was found while debugging #76729. fixing this issue would not fix this other one but would prevent the mouse from disappearing in this other issue.

### Steps to reproduce

- create a SpinBox
- add a script that in `_gui_input` wait for the mouse mode to be `MOUSE_MODE_CAPTURED` and then hide the spin box
- launch the game and do a drag modificiation on the `SpinBox`
- Congratulations your mouse is locked and you need to Alt + Tab to go back to the editor and force quit your game

### Minimal reproduction project

[test1.zip](https://github.com/godotengine/godot/files/11642625/test1.zip)

### PR DESCRIPTION
**Pull request by ajreckof**
Fixes #77803. Fixes https://github.com/godotengine/godot/issues/76729.

### CODE REVIEW

**Comment by Sauermann** [comment]
Have you thought about using `_release_mouse();` instead of `drag.allowed = false;`?
`NOTIFICATION_EXIT_TREE` uses that one to reset the mouse cursor.

**Comment by ajreckof** [comment]
So to give a bit more in depth. I do use `_release_mouse` as I fallthrough to the `NOTIFICATION_EXIT_TREE` case but this only reverting when drag is enabled. Unfortunatly because of how `SpinBox` this is not enough. This is because of `drag.allowed` which is different from `drag.enabled` the second indicates mouse has been captured and `_release_mouse` will handle it but the first indicates that on `InputEventMouseMotion` the `SpinBox` can capture the mouse and do the drag. 
calling `_release_mouse` on hiding fix part of the issue in case drag as already started but in case it hasn't started and `drag.enabled` is still false it might still appear this is in fact exactly what happens in #76729. Another way instead of setting `drag.allowed` there is to call it inside `_release_mouse` but I wasn't sure it wouldn't create other bug.

**Comment by korypostma** [comment]
I was looking into this issue as well on another issue and I added a few more notification types for when a user alt-tabs or if the app loses focus while dragging and I forcibly call `_release_mouse` as well.  I think those could be added here for more robustness.  We can discuss on rocketchat and you can make it part of this PR, since it is the same region of code for the same issue.

**Comment by korypostma** [comment]
`NOTIFICATION_FOCUS_EXIT`, `NOTIFICATION_WM_WINDOW_FOCUS_OUT`, and `NOTIFICATION_APPLICATION_FOCUS_OUT`

**Comment by korypostma** [comment]
I just tested this PR and it is breaking SpinBox functionality now.  The mouse cursor is always shown and only the up/down arrows work, this PR essentially breaks the dragging feature altogether.

File: scene/gui/spin_box.cpp [hunk 1]
```diff
@@ -249,6 +249,9 @@ void SpinBox::_notification(int p_what) {
 			_update_text();
 		} break;
 
+		case NOTIFICATION_VISIBILITY_CHANGED:
+			drag.allowed = false;
```
Comment (KoBeWi) [inline scene/gui/spin_box.cpp:253]: Shouldn't this only happen when the SpinBox is made invisible?

**Comment by KoBeWi** [comment]
> I just tested this PR and it is breaking SpinBox functionality now. The mouse cursor is always shown and only the up/down arrows work, this PR essentially breaks the dragging feature altogether.

I can't reproduce this. The PR works correctly.

File: scene/gui/spin_box.cpp (hunk 1 above)
Comment (ajreckof) [reply scene/gui/spin_box.cpp:253]: Yeah but I'm not sure it is in fact better to check performance wise then to set it both time whatever the case if it is invisible it should already be set to false. I think checking if it is in tree will be more costly then just setting the value. But I might be thinking badly about it feel free to correct me if it is the case

Comment (KoBeWi) [reply scene/gui/spin_box.cpp:253]: `is_visible_in_tree()` is a simple bool check, the value is cached.

Comment (ajreckof) [reply scene/gui/spin_box.cpp:253]: Always thought reading and writing bool value was the same time but I might be wrong moreover here we will have one check when hiding before setting which means for it to be worth it the function needs to always be twice faster the setting the value which I felt it was most likely the case

Comment (KoBeWi) [reply scene/gui/spin_box.cpp:253]: Well, the code I suggested also prevents calling `_release_mouse()`.
Although after reading it again, it doesn't really matter if the check exists or not.

**Review by KoBeWi — APPROVED** [review]

**Comment by ajreckof** [comment]
So looked a bit and effectively I can't reproduce the problem with the mrp but I think what they pointed out is how even though this is a half fix for #76729 it is not a full fix as their is a strange behaviour since it is being hidden and shown again every time the value is changed. A proper fix of this other bug would be to make sure it does hide and show it again every time it is updated. This is a problem of a long loop of update that had unforeseen consequences

**Comment by YuriSizov** [comment]
> Another way instead of setting `drag.allowed` there is to call it inside `_release_mouse` but I wasn't sure it wouldn't create other bug.

I think you should do that, yes. The only other place that calls `_release_mouse()` also sets the flag to off, so you could clean all of that up and make it a mandatory step to `drag.allowed = false;`.

**Review by YuriSizov — APPROVED** [review]
So this does solve the immediate problem, though you can still make your mouse disappear if you do something like hide, await 1-2 seconds, show as the drag happens.

I think this is bordering on shooting yourself in a foot at this point, though. After a rebase and a quick code clean up this should be good to go.

And this does fix https://github.com/godotengine/godot/issues/76729 btw, the spinbox is lagging like hell due to the described behavior issue, but the mouse pointer remains visible.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**